modules/security/src/main/java/io/oasp/module/security/common/base/accesscontrol/AbstractAccessControlProvider.java

```diff
@@ -63,17 +67,19 @@ protected void initialize(AccessControlSchema config) {
    * graph}.
    *
    * @param group is the {@link AccessControlGroup} to check.
-   * @param visitedGroups is where the {@link AccessControlGroup} are collected to detect cycles.
+   * @param nodeCycle the {@link NodeCycle} used to detect cycles.
    */
-  protected void checkForCyclicDependencies(AccessControlGroup group, Set<AccessControlGroup> visitedGroups) {
+  protected void checkForCyclicDependencies(AccessControlGroup group, NodeCycle<AccessControlGroup> nodeCycle) {
 
-    boolean added = visitedGroups.add(group);
-    if (!added) {
-      // mmm NodeCycleException would be very helpful here...
-      throw new IllegalStateException("Cyclic inheritance of access control groups detected for " + group);
-    }
     for (AccessControlGroup inheritedGroup : group.getInherits()) {
-      checkForCyclicDependencies(inheritedGroup, visitedGroups);
+      List<AccessControlGroup> inverseCycle = nodeCycle.getInverseCycle();
```

Comment:
I am fine with your suggestion as long as you provide an update or patch with that change. I am just reusing code that solves exactly what we need to have here. If you use ```java.util.Stack``` you will IMHO have to write more code and documentation around and would more or less duplicate the reused solution. However, I do not care and I seem to be biased according to mmm usage.